pixy/consumer.go

```diff
@@ -0,0 +1,566 @@
+	}()
+
+	timeoutErr := ErrConsumerRequestTimeout(fmt.Errorf("<%s> timeout", tc.contextID))
+	timeoutResult := consumeResult{Err: timeoutErr}
+	for consumeReq := range tc.requestsCh {
+		requestAge := time.Now().UTC().Sub(consumeReq.timestamp)
+		ttl := tc.config.Consumer.LongPollingTimeout - requestAge
+		// The request has been waiting in the buffer for too long. If we
+		// reply with a fetched message, then there is a good chance that the
+		// client won't receive it due to the client HTTP timeout. Therefore
+		// we reject the request to avoid message loss.
+		if ttl <= 0 {
+			consumeReq.replyCh <- timeoutResult
+			continue
+		}
+
+		select {
+		case msg := <-tc.messagesCh:
+			consumeReq.replyCh <- consumeResult{Msg: msg}
+		case <-time.After(ttl):
+			consumeReq.replyCh <- timeoutResult
+		}
+	}
+}
+
+func (tc *topicConsumer) String() string {
+	return tc.contextID.String()
+}
+
+// exclusiveConsumer ensures exclusive consumption of message from a particular
+// `consumer group+topic+partition`. It achieves that by claiming the partition
+// via the consumer group registry and only after the claim is succeeded starts
+// consuming messages.
+type exclusiveConsumer struct {
+	contextID    *sarama.ContextID
+	config       *Config
+	tc           *topicConsumer
+	group        string
+	topic        string
+	partition    int32
+	dumbConsumer sarama.Consumer
+	registry     *consumerGroupRegistry
+	offsetMgr    sarama.OffsetManager
+	stoppingCh   chan none
+	wg           sync.WaitGroup
+}
+
+func (cg *consumerGroup) spawnExclusiveConsumer(tc *topicConsumer, partition int32) *exclusiveConsumer {
+	ec := &exclusiveConsumer{
+		contextID:    cg.baseCID.NewChild(fmt.Sprintf("%s:%d", tc.topic, partition)),
+		config:       cg.config,
+		tc:           tc,
+		group:        cg.group,
+		topic:        tc.topic,
+		partition:    partition,
+		dumbConsumer: cg.dumbConsumer,
+		registry:     cg.registry,
+		offsetMgr:    cg.offsetMgr,
+		stoppingCh:   make(chan none),
+	}
+	spawn(&ec.wg, ec.run)
+	return ec
+}
+
+func (ec *exclusiveConsumer) run() {
+	defer ec.contextID.LogScope()()
+	defer ec.registry.claimPartition(ec.contextID, ec.topic, ec.partition, ec.stoppingCh)()
+
+	pom, err := ec.offsetMgr.ManagePartition(ec.group, ec.topic, ec.partition)
+	if err != nil {
+		log.Errorf("<%s> failed to spawn partition manager: err=(%s)", ec.contextID, err)
+		return
+	}
+	defer pom.Close()
+
+	var initialOffset sarama.FetchedOffset
+	for {
+		select {
+		case initialOffset = <-pom.InitialOffset():
+			goto spawnPartitionConsumer
+		case <-ec.stoppingCh:
+			return
+		}
+	}
+spawnPartitionConsumer:
+	pc, err := ec.dumbConsumer.ConsumePartition(ec.topic, ec.partition, initialOffset.Offset)
+	if err != nil {
+		log.Errorf("<%s> failed to start partition consumer: err=(%s)", ec.contextID, err)
+		return
+	}
+	defer pc.Close()
+
+	log.Infof("<%s> initialized: offset=%d", ec.contextID, initialOffset.Offset)
+	firstMessageFetched := true
```

Comment:
Somewhat minor, but given the name I would expect this to start false and then get flipped to true, rather than vice versa.